src/opers.c

```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
-              cache = 1+ADDR_OBJ( cacheBag );
-              cache[2*STATE(CacheIndex)] = method;
-              cache[2*STATE(CacheIndex)+1] = prec;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
-            }
-#ifdef COUNT_OPERS
-          OperationMiss++;
-#endif
-        }
-      if ( !method )  {
-        ErrorQuit( "no method returned", 0L, 0L );
-      }
-      
-      /* call this method                                                    */
-      res = CALL_0ARGS( method );
-    }
-    while (res == TRY_NEXT_METHOD );
-
-    /* return the result                                                   */
-    return res;
-}
-
-
-/****************************************************************************
-**
-**  DoOperation1Args( <oper>, <a1> )
-*/
-Obj DoOperation1Args (
-    Obj                 oper,
-    Obj                 arg1 )
-{
-    Obj                 res;
-    Obj                 type1;
-    Obj                 id1;
-    Obj *               cache;
-    Obj                 method;
-    Int                 i;
-    Obj                 prec;
-
-    /* get the types of the arguments                                      */
-    type1 = TYPE_OBJ_FEO( arg1 );
-    id1 = ID_TYPE( type1 );
-
-    /* try to find an applicable method in the cache                       */
-    prec = INTOBJ_INT(-1);
-
-    do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
-
-      /* recalculate cache each pass, in case of GC     */
-      cache = 1+ADDR_OBJ( CacheOper( oper, 1 ) );
 
-      /* Up to CACHE_SIZE methods might be in the cache */
-      if (prec < INTOBJ_INT(CACHE_SIZE))
-        {
-          for (i = 0;  i < 3*CACHE_SIZE; i+= 3) {
-            if (  cache[i+1] == prec && cache[i+2] == id1 ) {
-              method = cache[i];
+static inline Obj __attribute__((always_inline))
+GetMethodCached(Obj oper, UInt n, Obj prec, Obj ids[])
+{
+    UInt  typematch;
+    Obj * cache;
+    Obj   method = 0;
+    UInt  i;
+
+    cache = 1 + ADDR_OBJ(CacheOper(oper, n));
+
+    /* Up to CACHE_SIZE methods might be in the cache */
+    if (prec < INTOBJ_INT(CACHE_SIZE)) {
+        /* This loop runs through those */
+        UInt target = (n + 2) * INT_INTOBJ(prec);
+        for (i = target; i < (n + 2) * CACHE_SIZE; i += n + 2) {
+            if (cache[i + 1] == prec) {
+                typematch = 1;
+                /* This loop runs over the arguments, should be compiled away
+                 */
+                for (UInt j = 0; j < n; j++) {
+                    if (cache[i + j + 2] != ids[j]) {
+                        typematch = 0;
+                        break;
+                    }
+                }
+                if (typematch) {
+                    method = cache[i];
 #ifdef COUNT_OPERS
-              OperationHit++;
+                    CacheHitStatistics[INT_INTOBJ(prec)][i / (n + 2)][n]++;
```

Comment:
This function almost always uses `n+2`, and only uses `n` twice. So perhaps add an `n+=2;` at the start, and adjust accordingly?

Then `target` would become `const target = n * prec;`